commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +41,595 @@ type Client interface {
+	}
+
+	return signedObservations, nil
+}
+
+func (c *client) sendObservationRequests(
+	destChain *rmnpb.LaneDest,
+	requestsPerNode map[uint32][]*rmnpb.FixedDestLaneUpdateRequest,
+) (requestIDs mapset.Set[uint64]) {
+	requestIDs = mapset.NewSet[uint64]()
+
+	for nodeID, requests := range requestsPerNode {
+		req := &rmnpb.Request{
+			RequestId: newRequestID(),
+			Request: &rmnpb.Request_ObservationRequest{
+				ObservationRequest: &rmnpb.ObservationRequest{
+					LaneDest:                    destChain,
+					FixedDestLaneUpdateRequests: requests,
+				},
+			},
+		}
+
+		c.lggr.Infow("sending observation request", "node", nodeID, "requestID", req.RequestId)
+		if err := c.marshalAndSend(req, nodeID); err != nil {
+			c.lggr.Errorw("failed to send observation request", "node_id", nodeID, "err", err)
+			continue
+		}
+		requestIDs.Add(req.RequestId)
+	}
+	return requestIDs
+}
+
+func (c *client) listenForRmnObservationResponses(
+	ctx context.Context,
+	destChain *rmnpb.LaneDest,
+	requestIDs mapset.Set[uint64],
+	lursPerChain map[uint64]updateRequestWithMeta,
+	requestedNodes map[uint64]mapset.Set[uint32],
+) ([]rmnSignedObservationWithMeta, error) {
+	c.lggr.Infow("Waiting for observation requests", "requestIDs", requestIDs.String())
+
+	respChan := c.rawRmnClient.Recv()
+
+	finishedRequestIDs := mapset.NewSet[uint64]()
+	rmnObservationResponses := make([]rmnSignedObservationWithMeta, 0)
+	merkleRootsCount := make(map[uint64]map[string]int) // sourceChain -> merkleRoot -> count
+
+	initialObservationRequestTimer := time.NewTimer(c.observationsInitialRequestTimerDuration)
+	defer initialObservationRequestTimer.Stop()
+	for {
+		select {
+		case resp := <-respChan:
+			parsedResp, err := c.parseResponse(&resp, requestIDs, finishedRequestIDs)
+			if err != nil {
+				c.lggr.Warnw("failed to parse RMN observation response", "err", err)
+				continue
+			}
+			finishedRequestIDs.Add(parsedResp.RequestId)
+
+			signedObs := parsedResp.GetSignedObservation()
+			if signedObs == nil {
+				c.lggr.Infof("RMN node returned an unexpected type of response: %+v", parsedResp.Response)
+			} else {
+				err := c.validateSignedObservationResponse(resp.RMNNodeID, lursPerChain, signedObs, destChain)
+				if err != nil {
+					c.lggr.Warnw("failed to validate signed observation response", "err", err)
+					continue
+				}
+
+				c.lggr.Infow("received signed observation",
+					"node", resp.RMNNodeID, "requestID", parsedResp.RequestId)
+
+				rmnObservationResponses = append(rmnObservationResponses, rmnSignedObservationWithMeta{
+					RMNNodeID:         resp.RMNNodeID,
+					SignedObservation: signedObs,
+				})
+				for _, lu := range signedObs.Observation.FixedDestLaneUpdates {
+					if _, exists := merkleRootsCount[lu.LaneSource.SourceChainSelector]; !exists {
+						merkleRootsCount[lu.LaneSource.SourceChainSelector] = make(map[string]int)
+					}
+					merkleRootsCount[lu.LaneSource.SourceChainSelector][cciptypes.Bytes(lu.Root).String()]++
```

Comment:
Ah right the protobufs don't have an explicit Bytes32 type... wonder if we should check that len(lu.Root) == 32?

Reply:
Not super future proof if we ever get > 32 byte hashes, but should be fine for now